### PR TITLE
Add shell.writeShortcutLink/readShortcutLink APIs

### LINKED ISSUE #4414: Add create shortcut API
In order to use notifications on Windows 8-8.1, we need to create a shortcut. Instead of having to use a native module like [`node-win-shortcut`](https://www.npmjs.com/package/node-win-shortcut), could Electron provide a method? [NW.js does](https://github.com/nwjs/nw.js/blob/00eaaee63f668eae4e65fb812ad3f28d3411393c/src/api/app/app.cc#L154); e.g. `gui.App.createShortcut(process.env.APPDATA + "\\Microsoft\\Windows\\Start Menu\\Programs\\node-webkit.lnk")`

### PR DESCRIPTION
**Pull request by zcbenz**
```markdown

### `shell.writeShortcutLink(shortcutPath[, operation], options)` _Windows_

* `shortcutPath` String
* `operation` String (optional) - Default is `create`, can be one of followings:
  * `create` - Creates a new shortcut, overwriting if necessary.
  * `update` - Updates specified properties only on an existing shortcut.
  * `replace` - Overwrites an existing shortcut, fails if the shortcut doesn't
    exist.
* `options` Object
  * `target` String - The target to launch from this shortcut.
  * `cwd` String (optional) - The target to launch from this shortcut. Default
    is empty.
  * `args` String (optional) - The arguments to be applied to `target` when
    launching from this shortcut. Default is empty.
  * `description` String (optional) - The description of the shortcut. Default
    is empty.
  * `icon` String (optional) - The path to the icon, can be a DLL or EXE. `icon`
    and `iconIndex` have to be set together. Default is empty, which uses the
    target's icon.
  * `iconIndex` Integer (optional) - The resource ID of icon when `icon` is a
    DLL or EXE. Default is 0.
  * `appUserModelId` String (optional) - The Application User Model ID. Default
    is empty.

Creates or updates a shortcut link at `shortcutPath`. On success `true` is
returned, otherwise `false` is returned.

### `shell.readShortcutLink(shortcutPath)` _Windows_

Resolves the shortcut link at `shortcutPath`, an object is returned with the
fields described in the `options` of `shell.writeShortcutLink`.

An exception will be thrown when any error happens.
```

Close #4414.

### CODE REVIEW

**Comment by jorangreef** [comment]
Thanks @zcbenz.

**Comment by KochiyaOcean** [comment]
Is the path of icon must be a exe or dll file?

**Comment by MarshallOfSound** [comment]
@KochiyaOcean That's what the docs say.  Personally I just point do

```js
{
  icon: process.execPath,
  iconIndex: 0
}
```

That makes the shortcut the same icon as your program

**Comment by KochiyaOcean** [comment]
@MarshallOfSound thanks, it works!

File: atom/common/api/atom_api_shell.cc [hunk 1]
```diff
@@ -30,6 +58,61 @@ bool OpenExternal(
...
+  int index;
+  if (options.Get("target", &path))
+    properties.set_target(path);
+  if (options.Get("cwd", &path))
+    properties.set_working_dir(path);
+  if (options.Get("args", &str))
+    properties.set_arguments(str);
+  if (options.Get("description", &str))
+    properties.set_description(str);
+  if (options.Get("icon", &path) && options.Get("iconIndex", &index))
```
Comment (MarshallOfSound) [inline atom/common/api/atom_api_shell.cc]: @zcbenz The docs indicate that `iconIndex` has a default of 0 but this appears to not set the icon if `iconIndex` is not set.

Which behavior is expected / wanted and I will fix whichever end is wrong 😄

**Comment by paulcbetts** [comment]
Remember that Squirrel.Windows always writes its shortcut targets as trampolines - i.e. it's not a shortcut to `MyApp.exe`, it's a shortcut to `Update.exe --processStart MyApp.exe` -  this means that even if an update fails to fix the shortcut, it still is clickable.

https://github.com/Squirrel/Squirrel.Windows/blob/cedf4c296c830fcea09ff4ffca7cc1083b4ede67/src/Squirrel/UpdateManager.ApplyReleases.cs#L228

**Comment by MarshallOfSound** [comment]
@paulcbetts Yeah, I just updated my shortcut code to be something like this

```js
import { app, shell } from 'electron';
import fs from 'fs';
import path from 'path';

const packageJSON = require('./package.json');

const startPath = path.resolve(
  app.getPath('appData'), '..', 'Roaming', 'Microsoft', 'Windows', 'Start Menu', 'Programs',
  packageJSON.author.name, `${packageJSON.productName}.lnk`
);

const desktopPath = path.resolve(
  app.getPath('home'), 'Desktop', `${packageJSON.productName}.lnk`
);

const squirrelPath = path.resolve(path.dirname(process.execPath), '..', 'update.exe');

const shortcutAtPath = (targetPath) => {
  shell.writeShortcutLink(targetPath, fs.existsSync(targetPath) ? 'update' : 'create', {
    target: squirrelPath,
    args: `--processStart "${path.basename(process.execPath)}"`,
    icon: process.execPath,
    iconIndex: 0,
    appUserModelId: ID,
  });
};

export const createShortcut = () => {
  shortcutAtPath(desktopPath);
  shortcutAtPath(startPath);
};

export const removeShortcut = () => {
  if (fs.existsSync(startPath)) fs.unlinkSync(startPath);
  if (fs.existsSync(desktopPath)) fs.unlinkSync(desktopPath);
};
```

Gotta use them trampolines 😄

**Comment by KochiyaOcean** [comment]
Seems this api cant access icons in the asar achive :-(

**Comment by MarshallOfSound** [comment]
@KochiyaOcean That is correct, although you could at runtime copy the icons out of the ASAR onto the `appData` section of the users computer.

**Comment by m4heshd** [comment]
This doesn't seem to work anymore. Tried every possible way. It just doesn't create the shortcut. Return is also false.
OS - Windows 10.0.16299
FYI, `windows-shortcuts` worked. Unfortunately it doesn't have the **appUserModelId** option. I really wish this worked.

**Comment by Sivasss94** [comment]
@paulcbetts I have tried the above Shortcut code , but i while running i got the below mentioned issue.
untitled

**Pull request merged**